bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BTupleType.java

```diff
@@ -144,7 +144,17 @@ public boolean equals(Object o) {
             return false;
         }
         BTupleType that = (BTupleType) o;
-        return this.readonly == that.readonly && Objects.equals(tupleTypes, that.tupleTypes);
+
+        if ((this.restType == null || that.restType == null) && this.restType != that.restType) {
+            return false;
+        }
+
+        if (this.restType == null) {
```

Comment:
Shall we refactor here differently to take common branches out. I think it is easy to understand when should two tuple types should be equal that way.

eg:
```Java
if (this.readonly != that.readonly) {
  return false;
}

if (!Objects.equals(tupleTypes, that.tupleTypes)) {
  return false;
 }

if (this.restType != null) {
this.restType.equals(that.restType);
}

```

Reply:
Moved the read-only check to the top. But for the rest vs individual member comparisons, I think it'll be better to handle and be done with the rest scenario first, because it's either null or a comparison of a single pair of types. WDYT?